### ISSUE TITLE
[Train] Invalid checkpoints can result from some workers failing before finishing their checkpointing

### ISSUE THREAD

**Issue by justinvyu** [ray-project/ray#41700]
### What happened + What you expected to happen

How an invalid checkpoint can be created:
1. 2 workers A and B. Worker A finishes uploading checkpoint 0. Worker B dies before getting to finishing a checkpoint (ex: node failure)
2. At this point, storage contains the partially uploaded checkpoint 0 which contains A's contents but not B's contents.
3. However, Ray Train has not yet "recorded" a finished checkpoint 0, so the training run will restore from scratch.
4. When restored, workers A and B will both upload checkpoint 0 again. At this point, there may be contents from the previous (partial) checkpoint still.
5. **Note that this is not the biggest issue (P2) for checkpoints created by frameworks like lightning/deepspeed, since those have standard filenames, so the restored run will just overwrite the old files on upload.*
6. This also mostly applies to distributed checkpointing (where the concept of a partial checkpoint from a subset of workers is more likely than a partially uploaded checkpoint from a single worker). But, it can also happen for a single worker that has non-deterministic checkpoint contents based on iteration.
    * For example, if the checkpoint has a filename `ckpt-loss=0.123.pt`, where the loss is some non-deterministic value, then you may end up with two checkpoints within the same checkpoint directory on storage.

### Versions / Dependencies

2.7, 2.8, 2.9

### Reproduction script

```python
from pathlib import Path
import random
import tempfile

import ray
from ray.train.data_parallel_trainer import DataParallelTrainer
from ray.train import ScalingConfig, RunConfig, FailureConfig


def f(config):
    if ray.train.get_context().get_world_rank() == 0:
        marker: Path = config["marker"]
        first_run = marker.exists()
        marker.unlink(missing_ok=True)

    if ray.train.get_context().get_world_rank() == 0 and first_run:
        raise RuntimeError("worker 0 failing...")
    else:
        with tempfile.TemporaryDirectory() as temp_checkpoint_dir:
            rand = random.random()
            Path(temp_checkpoint_dir).joinpath(
                f"rank={ray.train.get_context().get_world_rank()}-rand={rand:.3f}"
            ).touch()
            ray.train.report(
                {"iter": 0},
                checkpoint=ray.train.Checkpoint.from_directory(temp_checkpoint_dir),
            )


marker = Path("/tmp/ray_results/marker")
marker.parent.mkdir(exist_ok=True)
marker.touch()

trainer = DataParallelTrainer(
    f,
    train_loop_config={"marker": marker},
    scaling_config=ScalingConfig(num_workers=4),
    run_config=RunConfig(
        failure_config=FailureConfig(max_failures=1), storage_path="/tmp/ray_results"
    ),
)
result = trainer.fit()
```

```
$ ls /path/to/checkpoint_000000
/tmp/ray_results/DataParallelTrainer_2023-12-07_10-26-48/DataParallelTrainer_304a5_00000_0_2023-12-07_10-26-51/checkpoint_000000
├── rank=0-rand=0.374
├── rank=1-rand=0.307
├── rank=1-rand=0.803
├── rank=2-rand=0.434
├── rank=2-rand=0.679
├── rank=3-rand=0.240
└── rank=3-rand=0.419
```

There are 8 files in here when there should only be 4 (4 workers).

### Issue Severity

Medium: It is a significant difficulty but I can work around it.
